### PR TITLE
fxtest: Fail test if app creation fails

### LINKED ISSUE #636: fxtest.New doesn't fail if the app isn't created successfully
```
func TestFxTestBug(t *testing.T) {
	app := fxtest.New(t,
		fx.Invoke(func(s string) {
			fmt.Println("invoked with", s)
		}),
	)
	fmt.Println("app error", app.Err())
}
```

Produces:
```
=== RUN   TestFxTestBug
app error missing dependencies for function "code.uber.internal/infra/uns.git/net/serverfx/serverfxtest".TestFxTestBug.func1 (/home/prashant/go/src/code.uber.internal/infra/uns.git/net/serverfx/serverfxtest/serverfxtest_test.go:102): type string is not in the container, did you mean to Provide it?
--- PASS: TestFxTestBug (0.00s)
	fxtest.go:43: [Fx] PROVIDE	fx.Lifecycle <= vendor/go.uber.org/fx.New.func1()
	fxtest.go:43: [Fx] INVOKE		code.uber.internal/infra/uns.git/net/serverfx/serverfxtest.TestFxTestBug.func1()
	fxtest.go:43: [Fx] Error during "code.uber.internal/infra/uns.git/net/serverfx/serverfxtest.TestFxTestBug.func1()" invoke: missing dependencies for function "code.uber.internal/infra/uns.git/net/serverfx/serverfxtest".TestFxTestBug.func1 (/home/prashant/go/src/code.uber.internal/infra/uns.git/net/serverfx/serverfxtest/serverfxtest_test.go:102): type string is not in the container, did you mean to Provide it?
PASS

Process finished with exit code 0
```

If I use `RequireStart` and `RequireStop`, it fails as expected, but I'd expect creation of the app to fail when using fxtest.

### PR DESCRIPTION
**Pull request by prashantv**
Currently, the fxtest.New function succeeds even if the underlying
fx.New fails. The error is never checked directly, and instead we rely
on `app.Start()` to catch the error. Instead, fail the test in
fxtest.New if fx.New fails.

There's some test changes required by this change:
 - The `StartFailure` test was actually a failure in New, so update test
   to fail in the `OnStart` hook
 - Some tests use fxtest.New, and then verify that the returned app has
   an error. they can use `New` directly, since `fxtest` doesn't any
   features on top apart from ensuring there's no errors.

Fixes #636

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,8 +5,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
-
-- No changes yet.
+### Fixed
+- Make `fxtest.New` ensure that the app was created successfully. Previously,
+  it would return the app (similar to `fx.New`, and expect the user to verify
+  the error.
```
Comment (abhinav) [inline CHANGELOG.md]: )

**Comment by akshayjshah** [comment]
> Good point, I've exported the TestPrinter from fxtest, and updated tests to use that.

Is there a commit hanging out on your machine? I don't see these changes in GitHub. (Sorry for the delay - I've been waiting for you to push changes...)

File: CHANGELOG.md (hunk 1 above)
Comment (akshayjshah) [reply CHANGELOG.md]: Please add trailing parens before landing.

File: fxtest/printer.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+// NewTestPrinter returns a fx.Printer that logs to the testing TB.
+func NewTestPrinter(t TB) fx.Printer {
+	return &testPrinter{t}
+}
+
+func (p *testPrinter) Printf(format string, args ...interface{}) {
+	p.Logf(format, args...)
+}
```
Comment (akshayjshah) [inline fxtest/printer.go]: Mind adding a trailing newline before landing? (For Unix-friendliness.)

**Pull request merged**